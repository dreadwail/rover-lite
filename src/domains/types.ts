import { Action, AnyAction } from 'redux';
import { ThunkAction, ThunkDispatch } from 'redux-thunk';

import { RoverSearchState } from '../state/types';

export type Omit<T, K extends keyof T> = Pick<T, Exclude<keyof T, K>>;

export type GenericThunkAction<T, S extends RoverSearchState = RoverSearchState> = ThunkAction<
  T,
  S,
  {},
  AnyAction
>;
export type PromiseThunkAction<
  T,
  S extends RoverSearchState = RoverSearchState
> = GenericThunkAction<Promise<T>, S>;

export type MapDispatchToProps<
  TDispatchProps,
  TOwnProps = {},
  TState extends RoverSearchState = RoverSearchState,
  A extends Action = Action
> = <E>(dispatch: ThunkDispatch<TState, E, A>, ownProps: TOwnProps) => TDispatchProps;
