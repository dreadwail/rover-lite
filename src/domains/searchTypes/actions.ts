import { Action } from 'redux';

import { request } from '../../api';
import { Endpoints } from '../../state/searchTypes/types';
import { PromiseThunkAction } from '../types';

export interface FetchSearchTypesPendingAction extends Action {
  readonly type: 'FETCH_SEARCH_TYPES.PENDING';
}

export interface FetchSearchTypesFulfilledAction extends Action {
  readonly type: 'FETCH_SEARCH_TYPES.FULFILLED';
  readonly payload: Endpoints;
}

export interface FetchSearchTypesRejectedAction extends Action {
  readonly type: 'FETCH_SEARCH_TYPES.REJECTED';
  readonly payload: {
    readonly error: any;
  };
  readonly error: true;
}

const fetchSearchTypesPending = (): FetchSearchTypesPendingAction => ({
  type: 'FETCH_SEARCH_TYPES.PENDING',
});

const fetchSearchTypesFulfilled = (endpoints: Endpoints): FetchSearchTypesFulfilledAction => ({
  type: 'FETCH_SEARCH_TYPES.FULFILLED',
  payload: endpoints,
});

const fetchSearchTypesRejected = (error: any): FetchSearchTypesRejectedAction => ({
  type: 'FETCH_SEARCH_TYPES.REJECTED',
  payload: {
    error,
  },
  error: true,
});

export type FetchSearchTypes = () => PromiseThunkAction<void>;
export const fetchSearchTypes: FetchSearchTypes = () => async dispatch => {
  dispatch(fetchSearchTypesPending());
  try {
    const response = await request<Endpoints>(
      'GET',
      'https://www.rover.com/api/v3/search-endpoints/'
    );
    dispatch(fetchSearchTypesFulfilled(response));
  } catch (err) {
    dispatch(fetchSearchTypesRejected(err));
  }
};
