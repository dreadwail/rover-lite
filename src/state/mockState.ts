import { RoverSearchState } from './types';

type Primitive = boolean | number | string;

export type RecursivePartial<ThisLevel> = {
  [Key in keyof ThisLevel]?: ThisLevel[Key] extends Primitive
    ? ThisLevel[Key]
    : RecursivePartial<ThisLevel[Key]>
};

export const createMockState = (
  interestingStateKeys?: RecursivePartial<RoverSearchState>
): RoverSearchState => (interestingStateKeys as unknown) as RoverSearchState;
