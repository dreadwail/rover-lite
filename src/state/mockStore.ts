import { AnyAction } from 'redux';
import configureStore, { MockStoreEnhanced } from 'redux-mock-store';
import thunkMiddleware, { ThunkDispatch } from 'redux-thunk';

export type MockStore<T> = MockStoreEnhanced<T, ThunkDispatch<T, {}, AnyAction>>;

export const createMockStore = <T>(initialState?: Partial<T>) => {
  const createStore = configureStore<T, ThunkDispatch<T, {}, AnyAction>>([thunkMiddleware]);
  if (initialState == null) {
    return createStore();
  }
  return createStore(initialState as T);
};
