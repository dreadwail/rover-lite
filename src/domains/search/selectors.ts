import { SearchInputs } from '../../state/search/types';
import { SearchType } from '../../state/searchTypes/types';
import { RoverSearchState } from '../../state/types';

export const getSearchType = (state: RoverSearchState): SearchType => state.search.searchType;
export const getInputs = (state: RoverSearchState): SearchInputs => state.search.inputs;
export const getLatLong = (state: RoverSearchState): [number, number] => [
  state.search.latitude,
  state.search.longitude,
];
