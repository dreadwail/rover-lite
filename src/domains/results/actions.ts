import { Action } from 'redux';

import { request } from '../../api/http';
import { urlEncodeQuery } from '../../api/query';
import { Listing } from '../../state/results/types';
import { SearchInputs } from '../../state/search/types';
import debouncedAction from '../debouncedThunk';
import { getInputs, getLatLong, getSearchType } from '../search/selectors';
import { PromiseThunkAction } from '../types';

import { normalize, RawSearchResponse } from './normalize';

export interface FetchResultsPendingAction extends Action {
  readonly type: 'FETCH_RESULTS.PENDING';
}

export interface FetchResultsFulfilledAction extends Action {
  readonly type: 'FETCH_RESULTS.FULFILLED';
  readonly payload: Listing[];
}

export interface FetchResultsRejectedAction extends Action {
  readonly type: 'FETCH_RESULTS.REJECTED';
  readonly payload: {
    readonly error: any;
  };
  readonly error: true;
}

const fetchResultsPending = (): FetchResultsPendingAction => ({
  type: 'FETCH_RESULTS.PENDING',
});

const fetchResultsFulfilled = (listings: Listing[]): FetchResultsFulfilledAction => ({
  type: 'FETCH_RESULTS.FULFILLED',
  payload: listings,
});

const fetchResultsRejected = (error: any): FetchResultsRejectedAction => ({
  type: 'FETCH_RESULTS.REJECTED',
  payload: {
    error,
  },
  error: true,
});

export type FetchResults = () => PromiseThunkAction<void>;
export const fetchResults: FetchResults = () => async (dispatch, getState) => {
  dispatch(fetchResultsPending());
  try {
    const state = getState();

    const searchType = getSearchType(state);
    const [latitude, longitude] = getLatLong(state);

    const defaultInputs: SearchInputs = {
      centerlat: latitude,
      centerlng: longitude,
      service_type: searchType,
    };
    const userInputs = getInputs(state);
    const inputs = { ...defaultInputs, ...userInputs };
    const encodedQuery = urlEncodeQuery(inputs);

    const urlBase = 'https://www.rover.com/api/v3/search';
    const url = `${urlBase}/${searchType}/?${encodedQuery}`;

    const response = await request<RawSearchResponse>('GET', url);

    const normalizedListings = normalize(response);
    dispatch(fetchResultsFulfilled(normalizedListings));
  } catch (err) {
    dispatch(fetchResultsRejected(err));
  }
};

export const debouncedFetchResults = debouncedAction(fetchResults, 250);
