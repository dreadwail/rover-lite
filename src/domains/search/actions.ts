import { Action } from 'redux';

import { SearchType } from '../../state/searchTypes/types';
import { GenericThunkAction } from '../types';

import { getInputs } from './selectors';

export interface SearchTypeChangedAction extends Action {
  readonly type: 'SEARCH.TYPE_CHANGED';
  readonly payload: SearchType;
}

export interface SearchFieldChangedAction extends Action {
  readonly type: 'SEARCH.FIELD_CHANGED';
  readonly payload: {
    readonly key: string;
    readonly value: any;
  };
}

export const searchTypeChanged = (searchType: SearchType): SearchTypeChangedAction => ({
  type: 'SEARCH.TYPE_CHANGED',
  payload: searchType,
});

export type SearchFieldChanged = (key: string, value: any) => GenericThunkAction<void>;
export const searchFieldChanged: SearchFieldChanged = (key, value) => async (
  dispatch,
  getState
) => {
  const state = getState();
  const inputs = getInputs(state);
  const startDate = inputs.start_date;
  const endDate = inputs.end_date;
  if (startDate != null && key === 'end_date' && startDate > value) {
    return;
  }
  if (endDate != null && key === 'start_date' && value > endDate) {
    return;
  }
  const action: SearchFieldChangedAction = {
    type: 'SEARCH.FIELD_CHANGED',
    payload: { key, value },
  };
  dispatch(action);
};
