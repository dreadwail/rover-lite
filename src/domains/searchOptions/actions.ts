import { Action } from 'redux';

import { request } from '../../api';
import { Options } from '../../state/searchOptions/types';
import { SearchType } from '../../state/searchTypes/types';
import { getEndpointUrl } from '../searchTypes/selectors';
import { PromiseThunkAction } from '../types';

import { normalize, OptionsResponse } from './normalize';

export interface FetchSearchOptionsPendingAction extends Action {
  readonly type: 'FETCH_SEARCH_OPTIONS.PENDING';
  readonly payload: {
    readonly searchType: SearchType;
  };
}

export interface FetchSearchOptionsFulfilledAction extends Action {
  readonly type: 'FETCH_SEARCH_OPTIONS.FULFILLED';
  readonly payload: {
    readonly searchType: SearchType;
    readonly options: Options;
  };
}

export interface FetchSearchOptionsRejectedAction extends Action {
  readonly type: 'FETCH_SEARCH_OPTIONS.REJECTED';
  readonly payload: {
    readonly searchType: SearchType;
    readonly error: any;
  };
  readonly error: true;
}

const fetchSearchOptionsPending = (searchType: SearchType): FetchSearchOptionsPendingAction => ({
  type: 'FETCH_SEARCH_OPTIONS.PENDING',
  payload: {
    searchType,
  },
});

const fetchSearchOptionsFulfilled = (
  searchType: SearchType,
  options: Options
): FetchSearchOptionsFulfilledAction => ({
  type: 'FETCH_SEARCH_OPTIONS.FULFILLED',
  payload: {
    searchType,
    options,
  },
});

const fetchSearchOptionsRejected = (
  searchType: SearchType,
  error: any
): FetchSearchOptionsRejectedAction => ({
  type: 'FETCH_SEARCH_OPTIONS.REJECTED',
  payload: {
    searchType,
    error,
  },
  error: true,
});

export type FetchSearchOptions = (searchType: SearchType) => PromiseThunkAction<void>;
export const fetchSearchOptions: FetchSearchOptions = searchType => async (dispatch, getState) => {
  dispatch(fetchSearchOptionsPending(searchType));
  try {
    const state = getState();
    const endpointUrl = getEndpointUrl(state, searchType);
    if (!endpointUrl) {
      throw new Error(`Unhandled search type: ${searchType}`);
    }
    const response = await request<OptionsResponse>('OPTIONS', endpointUrl);
    const normalizedOptions = normalize(response);
    dispatch(fetchSearchOptionsFulfilled(searchType, normalizedOptions));
  } catch (err) {
    dispatch(fetchSearchOptionsRejected(searchType, err));
  }
};
