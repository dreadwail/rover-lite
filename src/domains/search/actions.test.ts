import { createMockState } from '../../state/mockState';
import { createMockStore, MockStore } from '../../state/mockStore';
import { Search } from '../../state/search/types';
import { RoverSearchState } from '../../state/types';

import { searchFieldChanged, searchTypeChanged } from './actions';

describe('search actions', () => {
  describe('searchTypeChanged', () => {
    it('creates the correct action', () => {
      expect(searchTypeChanged('overnight-boarding')).toMatchSnapshot();
    });
  });

  describe('searchFieldChanged', () => {
    let state: RoverSearchState;
    let store: MockStore<RoverSearchState>;

    beforeEach(() => {
      const search: Search = {
        searchType: 'overnight-boarding',
        latitude: 47.6843725,
        longitude: -122.3874837,
        inputs: {},
      };
      state = createMockState({ search });
      store = createMockStore<RoverSearchState>(state);
      store.dispatch(searchFieldChanged('test-key', 'test-value'));
    });

    it('dispatches the correct actions', () => {
      expect(store.getActions()).toMatchSnapshot();
    });

    describe('with a start_date already in the store', () => {
      beforeEach(() => {
        const search: Search = {
          searchType: 'overnight-boarding',
          latitude: 47.6843725,
          longitude: -122.3874837,
          inputs: {
            start_date: new Date(1984, 7, 28),
          },
        };
        state = createMockState({ search });
        store = createMockStore<RoverSearchState>(state);
      });

      describe('when trying to change end_date to be less than start_date', () => {
        beforeEach(() => {
          store.dispatch(searchFieldChanged('end_date', new Date(1970, 0, 1)));
        });

        it('dispatches no actions', () => {
          expect(store.getActions()).toEqual([]);
        });
      });

      describe('when trying to change end_date to be greater than start_date', () => {
        beforeEach(() => {
          store.dispatch(searchFieldChanged('end_date', new Date(2050, 0, 1)));
        });

        it('dispatches a field changed action', () => {
          expect(store.getActions()).toMatchSnapshot();
        });
      });
    });

    describe('with a end_date already in the store', () => {
      beforeEach(() => {
        const search: Search = {
          searchType: 'overnight-boarding',
          latitude: 47.6843725,
          longitude: -122.3874837,
          inputs: {
            end_date: new Date(1984, 7, 28),
          },
        };
        state = createMockState({ search });
        store = createMockStore<RoverSearchState>(state);
      });

      describe('when trying to change start_date to be greater than end_date', () => {
        beforeEach(() => {
          store.dispatch(searchFieldChanged('start_date', new Date(2050, 0, 1)));
        });

        it('dispatches no actions', () => {
          expect(store.getActions()).toEqual([]);
        });
      });

      describe('when trying to change start_date to be less than end_date', () => {
        beforeEach(() => {
          store.dispatch(searchFieldChanged('start_date', new Date(1970, 0, 1)));
        });

        it('dispatches a field changed action', () => {
          expect(store.getActions()).toMatchSnapshot();
        });
      });
    });
  });
});
