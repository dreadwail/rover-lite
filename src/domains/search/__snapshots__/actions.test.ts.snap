// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`search actions searchFieldChanged dispatches the correct actions 1`] = `
Array [
  Object {
    "payload": Object {
      "key": "test-key",
      "value": "test-value",
    },
    "type": "SEARCH.FIELD_CHANGED",
  },
]
`;

exports[`search actions searchFieldChanged with a end_date already in the store when trying to change start_date to be less than end_date dispatches a field changed action 1`] = `
Array [
  Object {
    "payload": Object {
      "key": "start_date",
      "value": 1970-01-01T08:00:00.000Z,
    },
    "type": "SEARCH.FIELD_CHANGED",
  },
]
`;

exports[`search actions searchFieldChanged with a start_date already in the store when trying to change end_date to be greater than start_date dispatches a field changed action 1`] = `
Array [
  Object {
    "payload": Object {
      "key": "end_date",
      "value": 2050-01-01T08:00:00.000Z,
    },
    "type": "SEARCH.FIELD_CHANGED",
  },
]
`;

exports[`search actions searchTypeChanged creates the correct action 1`] = `
Object {
  "payload": "overnight-boarding",
  "type": "SEARCH.TYPE_CHANGED",
}
`;
